core/arch/arm/plat-sunxi/main.c

```diff
@@ -0,0 +1,172 @@
+register_phys_mem(MEM_AREA_IO_SEC, GIC_BASE, CORE_MMU_DEVICE_SIZE);
+#endif
+
+#ifdef CONSOLE_UART_BASE
+register_phys_mem(MEM_AREA_IO_NSEC,
+		  CONSOLE_UART_BASE, SUNXI_UART_REG_SIZE);
+#endif
+
+#ifdef SUNXI_TZPC_BASE
+register_phys_mem(MEM_AREA_IO_SEC, SUNXI_TZPC_BASE, SUNXI_TZPC_REG_SIZE);
+#define REG_TZPC_SMTA_DECPORT0_STA_REG      (0x0004)
+#define REG_TZPC_SMTA_DECPORT0_SET_REG      (0x0008)
+#define REG_TZPC_SMTA_DECPORT0_CLR_REG      (0x000C)
+#define REG_TZPC_SMTA_DECPORT1_STA_REG      (0x0010)
+#define REG_TZPC_SMTA_DECPORT1_SET_REG      (0x0014)
+#define REG_TZPC_SMTA_DECPORT1_CLR_REG      (0x0018)
+#define REG_TZPC_SMTA_DECPORT2_STA_REG      (0x001c)
+#define REG_TZPC_SMTA_DECPORT2_SET_REG      (0x0020)
+#define REG_TZPC_SMTA_DECPORT2_CLR_REG      (0x0024)
+#endif
+
+#ifdef SUNXI_CPUCFG_BASE
+register_phys_mem(MEM_AREA_IO_SEC, SUNXI_CPUCFG_BASE, SUNXI_CPUCFG_REG_SIZE);
+#endif
+
+#ifdef SUNXI_PRCM_BASE
+register_phys_mem(MEM_AREA_IO_SEC, SUNXI_PRCM_BASE, SUNXI_PRCM_REG_SIZE);
+#endif
+
+static struct gic_data gic_data;
+static TEE_Result tzpc_init(void);
+
+static void main_fiq(void)
+{
+	panic();
+}
+
+static const struct thread_handlers handlers = {
+	.std_smc = tee_entry_std,
+	.fast_smc = tee_entry_fast,
+	.nintr = main_fiq,
+	.cpu_on = pm_panic,
+	.cpu_off = pm_panic,
+	.cpu_suspend = pm_panic,
+	.cpu_resume = pm_panic,
+	.system_off = pm_panic,
+	.system_reset = pm_panic,
+};
+
+static struct sunxi_uart_data console_data;
+
+const struct thread_handlers *generic_boot_get_handlers(void)
+{
+	return &handlers;
+}
+
+void console_init(void)
+{
+#ifdef CFG_SUN8I_H2_PLUS
+	sunxi_uart_init(&console_data, CONSOLE_UART_BASE);
+	register_serial_console(&console_data.chip);
+#endif
+}
+
+#ifdef SUNXI_TZPC_BASE
+static TEE_Result tzpc_init(void)
+{
+	vaddr_t tzpc;
+
+	tzpc = (vaddr_t)phys_to_virt(SUNXI_TZPC_BASE, MEM_AREA_IO_SEC);
+
+	DMSG("SMTA_DECPORT0=%x", read32(tzpc+REG_TZPC_SMTA_DECPORT0_STA_REG));
+	DMSG("SMTA_DECPORT1=%x", read32(tzpc+REG_TZPC_SMTA_DECPORT1_STA_REG));
+	DMSG("SMTA_DECPORT2=%x", read32(tzpc+REG_TZPC_SMTA_DECPORT2_STA_REG));
+
+	/* Allow all peripherals for normal world */
+	write32(0xbe, tzpc+REG_TZPC_SMTA_DECPORT0_SET_REG);
+	write32(0xff, tzpc+REG_TZPC_SMTA_DECPORT1_SET_REG);
+	write32(0x7f, tzpc+REG_TZPC_SMTA_DECPORT2_SET_REG);
+
+	DMSG("SMTA_DECPORT0=%x", read32(tzpc+REG_TZPC_SMTA_DECPORT0_STA_REG));
+	DMSG("SMTA_DECPORT1=%x", read32(tzpc+REG_TZPC_SMTA_DECPORT1_STA_REG));
+	DMSG("SMTA_DECPORT2=%x", read32(tzpc+REG_TZPC_SMTA_DECPORT2_STA_REG));
+
+	return TEE_SUCCESS;
+}
+driver_init(tzpc_init);
+#endif /* SUNXI_TZPC_BASE */
+
+void main_init_gic(void)
+{
+	vaddr_t gicc_base;
+	vaddr_t gicd_base;
+
+	gicc_base = core_mmu_get_va(GIC_BASE + GICC_OFFSET, MEM_AREA_IO_SEC);
+	gicd_base = core_mmu_get_va(GIC_BASE + GICD_OFFSET, MEM_AREA_IO_SEC);
+
+	if (!gicc_base || !gicd_base)
+		panic();
+
+	/* Initialize GIC */
+	gic_init(&gic_data, gicc_base, gicd_base);
+	itr_init(&gic_data.chip);
+
+	/* SGI15 as Group-0 */
```

Comment:
So the CPU specific GIC initialization is done as line 155-166 for CPU0 and like `gic_cpu_init()` for CPU1. If this is intended behavoir it should be explained a bit more clear.

Reply:
I think maybe I can try to remove some of the code from 155-166. Some of them seems to be done in gic_init already. I copy those code from the ancient sunxi porting. It was for Sunxi-A80. That's why I have some Cortex-A15 code at very beginning because Sunxi-A80 is a big.little of Cortex-A15 and Cortex-A7. I'll check the doc and see if those code really affects.